In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv('../../Data/tip.csv')
t = pd.read_csv('../../Data/Titanic-Dataset.csv')

print(df.shape, t.shape)

(244, 7) (891, 12)


Part A — tips.csv (the restaurant)

A1. The owner wants to know how generous each customer was, not just how much they left.

Make a new column called tip_percent — the tip as a percentage of the bill, rounded to 1 decimal place. Then show only total_bill, tip and tip_percent for the first 5 rows.

In [ ]:
df['tip_percent'] = (df['tip'] / df['total_bill'] * 100).round(1)
print(df[['total_bill', 'tip', 'tip_percent']].head())

   total_bill   tip  tip_percent
0       16.99  1.01          5.9
1       10.34  1.66         16.1
2       21.01  3.50         16.7
3       23.68  3.31         14.0
4       24.59  3.61         14.7


A2. Answer both parts using positions only — no column names anywhere.

Print the total_bill of the 10th row.
Show the last 3 rows, with only the first two columns.

In [3]:
print(df.iloc[9, 0])
print(df.iloc[-3:, :2])

14.78
     total_bill   tip
241       22.67  2.00
242       17.82  1.75
243       18.78  3.00


A3. Now do it with labels only.

Show rows 20 to 25, with all the columns from sex through size.

Then answer in a comment: how many rows came back, and why is it not 5? ⭐

In [4]:
print(df.loc[20:25, 'sex':'size'])

       sex smoker  day    time  size
20    Male     No  Sat  Dinner     2
21  Female     No  Sat  Dinner     2
22  Female     No  Sat  Dinner     2
23    Male     No  Sat  Dinner     4
24    Male     No  Sat  Dinner     2
25    Male     No  Sat  Dinner     4


A4. Two counts for the owner:

How many bills were not at Dinner? Use ~.
How many bills had a party of 3 or more and a tip above 4?

In [5]:
print((~(df['time'] == 'Dinner')).sum())
print(((df['size'] >= 3) & (df['tip'] > 4)).sum())

68
28


A5. The owner wants to flag the generous customers.

Make a column called service. Everyone starts as 'normal'. Any bill where tip_percent is more than 20 becomes 'generous'.

Then count how many of each. (This uses the column you made in A1.)

In [6]:
df['service'] = 'normal'
df.loc[df['tip_percent'] > 20, 'service'] = 'generous'
print(df['service'].value_counts())

service
normal      206
generous     38
Name: count, dtype: int64


Part B — Titanic-Dataset.csv

⚠️ Remember: the column names here start with capital letters — Survived, Pclass, Sex, Age, SibSp, Parch, Fare, Cabin, Embarked.

B1. How many passengers boarded from each port? Use the Embarked column.

Then add the three numbers up in a comment. Do they come to 891? ⭐

In [7]:
print(t['Embarked'].value_counts())

Embarked
S    644
C    168
Q     77
Name: count, dtype: int64


B2. Two counts:

How many women were travelling in third class?
How many of those women survived?

In [8]:
women_third = (t['Sex'] == 'female') & (t['Pclass'] == 3)
print(women_third.sum())
print((women_third & (t['Survived'] == 1)).sum())

144
72


B3. Some people paid an enormous amount for their ticket.

Show only Name, Age and Fare for every passenger who paid more than 200. How many are there?

In [9]:
expensive = t.loc[t['Fare'] > 200, ['Name', 'Age', 'Fare']]
print(expensive)
print(len(expensive))

                                                  Name   Age      Fare
27                      Fortune, Mr. Charles Alexander  19.0  263.0000
88                          Fortune, Miss. Mabel Helen  23.0  263.0000
118                           Baxter, Mr. Quigg Edmond  24.0  247.5208
258                                   Ward, Miss. Anna  35.0  512.3292
299    Baxter, Mrs. James (Helene DeLaudeniere Chaput)  50.0  247.5208
311                         Ryerson, Miss. Emily Borie  18.0  262.3750
341                     Fortune, Miss. Alice Elizabeth  24.0  263.0000
377                          Widener, Mr. Harry Elkins  27.0  211.5000
380                              Bidois, Miss. Rosalie  42.0  227.5250
438                                  Fortune, Mr. Mark  64.0  263.0000
527                                 Farthing, Mr. John   NaN  221.7792
557                                Robbins, Mr. Victor   NaN  227.5250
679                 Cardeza, Mr. Thomas Drake Martinez  36.0  512.3292
689   

B4. Out of all the passengers in first class, what percentage survived?

Now do the same for third class. Put the two numbers next to each other in a comment.

In [10]:
first_class = t[t['Pclass'] == 1]
third_class = t[t['Pclass'] == 3]
first_percentage = first_class['Survived'].mean() * 100
third_percentage = third_class['Survived'].mean() * 100
print(first_percentage)
print(third_percentage)

62.96296296296296
24.236252545824847


B5. Make a column called age_group. Everyone starts as 'adult'. Anyone under 18 becomes 'child'. Then count how many of each.

Think before tomorrow (no code needed). 177 passengers have no Age recorded at all. Which group did they land in — and is that the right place for them? Write two lines in a comment saying what you would do about it. ⭐

In [11]:
t['age_group'] = 'adult'
t.loc[t['Age'] < 18, 'age_group'] = 'child'
print(t['age_group'].value_counts())

age_group
adult    778
child    113
Name: count, dtype: int64
